stdlib/fun.mli

```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
     an unexpected exception or a programming error. As a general rule,
     one should not catch a [Finally_raised] exception except as part of
     a catch-all handler. *)
+
+(** {1:examples Examples}
+
+{2 Combinators}
+
+{{!combinators}Combinators} provide a lightweight and sometimes more readable
+way to create anonymous functions, best used as short-lived arguments rather
+than standalone definitions. The examples below will demonstrate this mainly
+with the {!module:List} module.
+
+
+{3:hid {{!val:id}id}}
+
+{!val:List.init} with the index itself
+{[
+# List.init 3 Fun.id;;
+- : int list = [0; 1; 2]
+]}
+
+Using {!val:List.filter_map} on an [int option list] to filter out
+{{!const:Option.t.None}[None]} elements
+{[
+# List.filter_map Fun.id [None; Some 2; Some 3; None; Some 5];;
+- : int list = [2; 3; 5]
+]}
+
+Using {!val:Float.Array.map_from_array} to flatten a [float array]
+{[
+# let to_flat arr = Float.Array.map_from_array Fun.id arr;;
+val to_flat : float array -> Float.Array.t
+]}
+
+Dispatching functions of type [foo -> foo] conditionally is another place where
+[id] may be useful. Consider a function which either uses {!val:String.map} on
+a path to convert unix-style slashes to windows-style backslashes or leaves it
+as it is depending on the {!val:Sys.win32} value
+{[
+if Sys.win32 then String.map (function '/' -> '\\' | c -> c) else Fun.id
+]}
+
+And for more advanced uses, where we may build up closures, [id] is often used
+for the base-case as a no-op. Consider a function which chains a list of unary
+functions:
+{[
+let rec chain = function
+  | [] -> Fun.id
+  | f :: fs -> fun x -> f (chain fs x)
+]}
```

Comment:
Another advanced example is as a final continuation for functions in continuation-passing-style:

```ocaml
type 'a btree =
  | Leaf of 'a
  | Node of 'a btree * 'a * 'a btree

let rec map_cps f tree k =
  match tree with
  | Leaf v -> k (Leaf (f v))
  | Node (left, v, right) ->
    map_cps f left @@ fun left' ->
    let v' = f v in
    map_cps f right @@ fun right' ->
    Node (left', v', right')

let map f tree = map_cps f tree Fun.id
```

Reply:
That was originally in place of `chain`, but I wasn't sure if introducing the concept of CPS in an example was a case worth defending